# Contents
01 Importing Libraries
02 Data Wrangling
03 Data Quality
04 Combining Data Sets

# 01 Importing Libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Folder location string
path = r'C:\Users\ckra9\Documents\CareerFoundry\Instacart Basket Analysis'

In [3]:
# Importing customer data set
df_customer = pd.read_csv(os.path.join(path, '02 Data','Original Data', 'customers.csv'))

In [4]:
df_customer.shape

(206209, 10)

In [4]:
df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# 02 Q4 Data wrangling

In [5]:
df_customer.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [6]:
# renaming columns
df_customer2 = df_customer.rename(columns = {'First Name': 'first_name', 'Surnam': 'last_name', 'Gender': 'gender', 'STATE': 'state', 'Age': 'age'})

In [7]:
df_customer2.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# 03 Q5 Data quality

In [8]:
# Checking for mixed data types
for col in df_customer2.columns.tolist():
  weird = (df_customer2[[col]].applymap(type) != df_customer2[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer2[weird]) > 0:
    print (col)

first_name


In [9]:
# bringing up mixed data type instances in the first_name column & checking for missing values throughout the dataframe
df_customer2.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

There are 11,259 instances of missing values in the first_name column

In [10]:
df_first_name_missing = df_customer2[df_customer2['first_name'].isnull() == True]

In [11]:
df_first_name_missing

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

In [12]:
# Imputing missing first name with Unknown
df_customer2['first_name'].fillna('Unknown', inplace=True)

In [13]:
# Checking work
df_customer2.isnull().sum()

user_id         0
first_name      0
last_name       0
gender          0
state           0
age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

In [14]:
# Creating dataframe containing duplicates from customer data
df_customer_dups = df_customer2[df_customer2.duplicated()]

In [15]:
df_customer_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

There doesn't appear to be any duplicates within the dataframe

# 04 Q6 Combining data sets

In [16]:
# Importing prepared instacart data
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merge_agg_flags.pkl'))

In [17]:
# Checking data type of user_id customer
df_customer2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   first_name    206209 non-null  object
 2   last_name     206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [18]:
# Checking data type of user_id orders products
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 21 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                object  
 3   order_number            int64   
 4   order_dow               int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  product_name            object  
 11  aisle_id                int64   
 12  department_id           int64   
 13  prices                  float64 
 14  _merge                  category
 15  max_order               int64   
 16  loyalty_flag            object  
 17  user_prices_mean        float64 
 18  spending_flag           object  
 19  median_days             float64 
 20  frequency_flag          object  
dtypes: cat

In [19]:
# Merging data sets on user id
ords_prods_cust_merge = df_customer2.merge(ords_prods_merge, on = 'user_id')

In [20]:
ords_prods_cust_merge.head()

user_id first_name last_name  gender     state  age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   n_dependants fam_status  income  ...  aisle_id department_id  prices  \
0             3    married  165665  ...        77             7     9.0   
1             3    married  165665  ...        77             7     9.0   
2             3    married  165665  ...        77             7     9.0   
3             3    married  165665  ...       120            16    12.6   
4             3    married  165665  ...       120            16    12.6   

   _merge  max_order  loyalty_flag  user_prices_mean  spending_flag  \
0    both          8  New customer          7.988889    Low spender   
1    both          8  New customer          7.988889    Low spender   
2    both          8  New customer          7.988889    Low spender   
3    both          8  New customer          7.988889    Low spender   
4    both          8  New customer          7.988889    Low spender   

   median_days    frequency_flag  
0         19.0  Regular customer  
1         19.0  Regular customer  
2         19.0  Regular customer  
3         19.0  Regular customer  
4         19.0  Regular customer  

[5 rows x 30 columns]

In [21]:
ords_prods_cust_merge.shape

(32404859, 30)

In [22]:
ords_prods_merge.shape

(32404859, 21)

In [23]:
df_customer2.shape

(206209, 10)

In [24]:
# Checking merge
ords_prods_cust_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [25]:
# Exporting merged dataframe
ords_prods_cust_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_cust_merge.pkl'))